File: C++/LeetCode/LeetCode/207_Course_Schedule.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;
class Solution {
public:
	bool dfs(int i, vector<int>& visited, unordered_map<int, vector<int>>& graph) {
		if (visited[i] == -1) {
			return false;
		}
		if (visited[i] == 1) {
			return true;
		}
		visited[i] = -1;
		for (auto vertex : graph[i]) {
			if (!dfs(vertex, visited, graph)) {
				return false;
			}
		}
		visited[i] = 1;
		return true;
	}

	bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
		unordered_map<int, vector<int>> graph;
		vector<int> visited(numCourses, 0);
		for (auto edge : prerequisites) {
			graph[edge[0]].push_back(edge[1]);
		}
		for (int i = 0; i < numCourses; i++) {
			if (!dfs(i, visited, graph)) {
				return false;
			}
		}
		return true;

	}
};

File: python/leetcode/207_Course_Schedule.py
# /usr/bin/python3.6
# -*- coding:utf-8 -*-


class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        graph = [[] for _ in range(numCourses)]
        visited = [0 for _ in range(numCourses)]
        for pre, n in prerequisites:
            graph[pre].append(n)

        def dfs(i):
            if visited[i] == -1:
                return False
            if visited[i] == 1:
                return True
            visited[i] = -1
            for j in graph[i]:
                if not dfs(j):
                    return False
            visited[i] = 1
            return True

        for i in range(numCourses):
            if not dfs(i):
                return False
        return True


def main():
    s = Solution()
    print(s.canFinish(2, [[0, 1]]))


if __name__ == "__main__":
    main()
